Remove current IBL resource if empty or incorrect URL provided.

amarchenko opened the pull request:
#### 2d152608cf6e96e894c616aa7702aa768b6c4365
<pre>
Remove current IBL resource if empty or incorrect URL provided.

<a href="https://bugs.webkit.org/show_bug.cgi?id=282182">https://bugs.webkit.org/show_bug.cgi?id=282182</a>
<a href="https://rdar.apple.com/137812845">rdar://137812845</a>

Reviewed by Ada Chan.

These changes make sure that an empty URL removes provided IBL lighting and
resolves environmentMapPromise. An invalid URL removes IBL lighting as well but
rejects environmentMapPromise.
Also, WKSRKEntity::applyIBL takes a completion handler to provide result of
the asynchronous call. The result is propagated all the way to HTMLModelElement
via didFinishEnvironmentMapLoading(bool).
Last but not least, when environment map resource/URL is provided before model
data is ready, the environment lighting will be applied to the model. That covers
a case when model resource is changed but environment map stays the same.

Canonical link: <a href="https://commits.webkit.org/285878@main">https://commits.webkit.org/285878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/381ca0c214b2c4657da8e5bf4a3a38b40dfccaef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78409 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76172 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1269 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58206 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16568 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48374 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63702 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38617 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45239 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21194 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23626 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21541 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79947 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/747 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66529 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65804 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16315 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9725 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7896 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1336 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1365 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1353 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -698,18 +705,39 @@ void HTMLModelElement::setEnvironmentMap(const URL& url)
+        if (m_modelPlayer)
+            m_modelPlayer->setEnvironmentMap(SharedBuffer::create());
+        return;
     }
```
Should this logic of checking for empty string go into `HTMLModelElement::selectEnvironmentMapURL()`? And have that method returns a `std::optional<URL>`? 

`HTMLModelElement::setEnvironmentMap()` can then take in an optional URL too, and remove the environment map if the URL is null?

amarchenko replied:
I agree that checking attribute emptiness should be moved to `selectEnvironmentMapURL` but we still need a check `m_environmentMapURL.isEmpty()` in `setEnvironmentMap` to avoid unnecessary resource requests.

As for std::optional<URL> I'd prefer not to use it here because a) for consistency sake (see model source URL), b) `URL()` covers empty case.

Updated code.

amarchenko committed:
Remove current IBL resource if empty or incorrect URL provided.

https://bugs.webkit.org/show_bug.cgi?id=282182
rdar://137812845

Reviewed by Ada Chan.

These changes make sure that an empty URL removes provided IBL lighting and
resolves environmentMapPromise. An invalid URL removes IBL lighting as well but
rejects environmentMapPromise.
Also, WKSRKEntity::applyIBL takes a completion handler to provide result of
the asynchronous call. The result is propagated all the way to HTMLModelElement
via didFinishEnvironmentMapLoading(bool).
Last but not least, when environment map resource/URL is provided before model
data is ready, the environment lighting will be applied to the model. That covers
a case when model resource is changed but environment map stays the same.

Canonical link: https://commits.webkit.org/285878@main
